### DISCUSSION TITLE
Validate document before inserting

### DISCUSSION THREAD

**Discussion by dv336699** [appwrite/appwrite#1587] [Q&A] [answered]
Is there a way to stop a document from being updated or created eg if one of the fields don’t match a regex?

Parse call this Triggers
https://docs.parseplatform.org/cloudcode/guide/#save-triggers

Thanks

**Comment by eldadfux** [comment, marked as answer]
You could achieve the same thing with the next release (0.13) which will introduce sync+async functions. Instead of letting the user create the document, you can delegate that to a custom function that can do whatever logic that you want and decide if to create the document or not.

**Reply by dv336699** [reply]
@eldadfux thanks for that.

A follow up question if that's ok.

Do I need to return an exception within the custom function to allow the client to catch it?

looking at these two examples 
- https://github.com/appwrite/demos-for-functions/blob/master/nodejs/add-created-at/index.js
- https://github.com/appwrite/demos-for-functions/blob/master/php/add-created-at/index.php

I can see that these return a string or an object - does that means that this will be returned to the client if I have a function that triggers when a document is being created?

thanks

**Reply by eldadfux** [reply]
In versions <= 0.12 we execute the functions in CLI mode. So you need to return CLI errors, for example in php, you would return the exit function with an exit code != 0. In 0.13 we will be upgrading the performance by moving to an http execution model over the CLI one (much much faster). In the new execution model your code could leverage exceptions natively without needing to wrap and translate them to exit codes.

Once, ready we will cover this issue on the official docs.
